text/0558-edition-detection.md

```diff
@@ -0,0 +1,136 @@
+- Start Date: (2019-11-20)
+- Relevant Team(s): (Ember.js, Ember CLI, Ember Data)
+- RFC PR: https://github.com/emberjs/rfcs/pull/558
+- Tracking: (leave this empty)
+
+# Edition detection
+
+## Summary
+
+Introduces a mechanism that an application can use to specify which Edition of
+Ember it is intending to target. This RFC will define:
+
+* How to specify the edition that the application is using
+* How other packages (addons, codemods, etc) can detect the applications intended edition
+* What the edition should be used for
+
+## Motivation
+
+As Ember approaches it's first edition (see
+[emberjs/rfcs#364](https://github.com/emberjs/rfcs/pull/364)) various addons
+need to modify their behavior based on the edition that is being used. An
+initial implementation (done without RFC) used the `setEdition` method from
+`@ember/edition-utils` inside the application's (or addon's) `.ember-cli.js`
+file to specify which edition was to be used. That implementation worked well
+enough throughout the intial preview period, but a number of major issues were
+(rightfully!) surfaced by the community:
+
+1. it seems unnecessary/redundant
+2. it's not clear what this flag actually does (likely due to having no RFC!)
+3. it's not statically analyzable (and therefore cannot be used by things like codemods)
+
+This RFC will review these issues in light of the updated implementation,
+showing how each of the concerns have been met.
+
+## Detailed design
+
+### Specifying the edition
+
+A new entry will be added to the project's `package.json`:
+
+```js
+{
+  "name": "project-name",
+  // ...snip
+  "ember": {
+    "edition": "octane"
+  }
+}
+```
+
+The value will be expected to be one of the valid edition names (currently
+`classic` and `octane`). Using the `package.json` for this allows us to ensure
+that the value is statically analyzable and easy to discover.
+
+### Valid use of the edition value
+
+The edition flag should only be used by addons to determine what blueprint
+output to generate and to provide helpful warnings (or errors) at build time.
+
+### Detecting the edition
+
+The existing `@ember/edition-utils` package will still be used by addons to
+detect which edition is in use, but it will be updated to check the new
+location (instead of relying on folks leveraging `setEdition`).
+
+The API documentation for `@ember/edition-utils` would be:
+
+```ts
+module '@ember/edition-utils' {
+  /**
+    Determine if the application that is running is running under a given Ember
+    Edition.  When the edition in use is _newer_ than the requested edition it
+    will return `true`, and if the edition in use by the application is _older_
+    than the requested edition it will return `false`.
```

Comment:
What, when we get the first edition, that is a major bump (4.x)? Would `octane` still return `true`, since the next edition is strictly newer?

Reply:
Yep. The idea is that calling `has('octane')` will be true if the app specifies any edition from octane and later. Without that functionality every addon that wants to do any work here would have to duplicate the logic to know which strings are newer and which are older (and we **definitely** don’t want to use version numbers)

Reply:
My only concern was future outdated addons that rely on features introduced in `octane` or implicitly rely on other features, that may be removed with 4.x, would break then. But I think this is a pretty minor hazard, if any at all.

Reply:
> My only concern was future outdated addons that rely on features introduced in octane or implicitly rely on other features, that may be removed with 4.x, would break then.

Ya, but that hazard is fairly unrelated to the edition. APIs removed in a major version bump would always have this pitfall (and is one of the reasons we have so few major bumps!).